DCOS-59714: Document application submission and troubleshooting (#39)
mesosphere/kudo-spark-operator: kudo-operator/docs/latest/submission.md

Submitting Spark Applications
---

> In order to deploy your application to Kubernetes using Kudo Spark Operator, first you need to describe it as a Kubernetes object. To do that, you need to create a specification in `yaml` format, that will contain all the configuration required for the application.

Reviewer: let's add a prerequisites section explicitly mentioning that KUDO Spark must be installed

> In order to deploy your application to Kubernetes using Kudo Spark Operator, first you need to describe it as a Kubernetes object. To do that, you need to create a specification in `yaml` format, that will contain all the configuration required for the application.

Reviewer: ```suggestion
In order to deploy a Spark Application to Kubernetes using the KUDO Spark Operator, it should be described as a Kubernetes object. To do that, create a specification in `yaml` format with all the necessary configuration required for the application.
```

```yaml
apiVersion: "sparkoperator.k8s.io/v1beta2"
kind: SparkApplication
metadata:
  name: spark-pi
  namespace: spark
spec:
  type: Scala
  mode: cluster
  image: "mesosphere/spark:spark-2.4.3-hadoop-2.9-k8s"
  imagePullPolicy: Always
  mainClass: org.apache.spark.examples.SparkPi
  mainApplicationFile: "local:///opt/spark/examples/jars/spark-examples_2.11-2.4.3.jar"
  arguments:
    - "150000"
  sparkConf:
    "spark.ui.port": "4041"
  sparkVersion: "2.4.3"
  restartPolicy:
    type: Never
  driver:
    cores: 1
    memory: "512m"
    labels:
      version: 2.4.3
    serviceAccount: spark-driver
  executor:
    cores: 1
    instances: 2
    memory: "512m"
    labels:
      version: 2.4.3
```

Basically, all the Spark application configuration is placed under `spec` section. Here you can specify Spark related configuration properties, such as number of executors, number of cores for drivers/executors, amount of memory and the other. There is also a `sparkConf` section, where you can place configuration parameters in a form of key-value pairs. In the example we override the default `spark.ui.port` with a custom value.

> After the application spec is ready, the following command can be used to run it on the operator:

Reviewer: ```suggestion
After the application spec is ready, the following command can be used to submit it to the operator:
```

```
$ kubectl create -f ./resources/spark-pi.yaml
```

> 

Reviewer: it makes sense to add a `list` command (`k get  sparkapplications -n spark`) to display all Spark Applications and verify that `spark-pi` is created before getting the specific one

> To describe the newly created application, you can use:

Reviewer: ```suggestion
To describe the newly created application, use the following command:
```

```bash
$ kubectl get sparkapplications.sparkoperator.k8s.io spark-pi -n spark
```

> Pay attentions to the `-n spark` flag - it the name of the namespace, where our application is deployed. 

Reviewer: ```suggestion
In the example above `-n` flag specifies the namespace where the application is deployed. 
```

To get the list of pods:
```bash
$ kubectl get pods -n spark
NAME                                  READY   STATUS      RESTARTS   AGE
spark-gq4bf8-spark-867b8c8d79-btt4q   1/1     Running     0          58m
spark-gq4bf8-spark-init-cdcrn         0/1     Completed   0          58m
spark-pi-1571911449587-exec-1         1/1     Running     0          4s
spark-pi-1571911449587-exec-2         1/1     Running     0          4s
spark-pi-driver                       1/1     Running     0          11s
```

> From the output above we can see the app has been created and is currently running as three pods: one driver and two executors.

Reviewer: ```suggestion
From the output above we can see the application has been created and is currently running as three pods: one driver and two executors.
```

To see pod's logs, use:
```
$ kubectl logs -f spark-pi-driver -n spark
```
Get the last `n` lines of pod's log:
```
$ kubectl logs --tail=n spark-pi-driver -n spark
```
You can get the full description of the driver pod with the following command: 
```
$ kubectl describe pod spark-pi-driver -n spark
```
Verify the driver pod is completed successfully:
```bash
$ kubectl logs --tail=20 spark-pi-driver -n spark | grep 'Pi is'
Pi is roughly 3.141644502283289
```

> 

Reviewer: It would be great to add instructions for application deletion and also for restarting the previously created application with updated parameters.

### Accessing Spark UI
When running Spark applications on K8s, it could be useful to have an access to Spark UI to monitor and inspect Spark job executions in a web browser. 

> *Tip: *

Reviewer: shall this be a header or removed completely? I think the explanation fits well as a continuation of "overview" and doesn't need any special treatment

The are a few ways to expose Spark UI for you application. 

> When an application is submitted to Kubernetes, the operator automatically creates a `Service` with default type `ClusterIP`, which can be used to access Spark UI externally.

Reviewer: do we have a parameter controlling automatic Service creation for DriverUI?

Author: I looked through the sources - it looks like there is no way to tweak that via configuration. Furthermore, a service is created even when `spark.ui.enabled` is set to `false` in `sparkConf` section, which looks like a bug btw

Reviewer: you're right, sorry for the confusion. The above doc is correct, Driver UI service is created automatically and this behaviour is not configurable. It happens here: https://github.com/mesosphere/spark-on-k8s-operator/blob/master/pkg/controller/sparkapplication/controller.go#L682

   
```bash
$ kubectl get svc -n spark --field-selector metadata.name=spark-pi-ui-svc
NAME                                  TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)             AGE
spark-pi-ui-svc                       ClusterIP   10.0.5.197    <none>        4041/TCP            7s
```

#### Using Port Forwarding

> Port forwarding works in a way that connections made to a local port are forwarded to port of the pod that is running the Spark driver. With this connection in place you can use your local workstation to access Spark UI that is running in the Driver pod.

Reviewer: ```suggestion
Port forwarding works in a way that connections made to a local port are forwarded to port of the pod which is running the Spark driver. With this connection in place, you can use your local workstation to access Spark UI which is running in the Driver pod.
```

Command example:
```
$ kubectl port-forward <resource-name> <local-port>:<container-port>
```

In case with `spark-pi` application, the command will look like the following:
```bash
$ kubectl port-forward spark-pi-driver 4040:4041 -n spark
Forwarding from 127.0.0.1:4040 -> 4041
Forwarding from [::1]:4040 -> 4041

```

> After that you can verify Spark UI is available on `localhost:4040`:

Reviewer: ```suggestion
After that the Spark UI should be available via URL: [localhost:4040](localhost:4040):
```

Service specification can look like the following:  
```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: spark-pi-loadbalancer
  name: spark-pi-loadbalancer
spec:
  type: LoadBalancer
  selector:
    spark-role: driver
  ports:
    - protocol: TCP
      port: 80
      targetPort: 4041
```

Specifying `spark-role: driver` as a `selector` is basically the mechanism for our service to target the required pod.

Create the service using `kubectl create -f ./resources/spark-pi-service.yaml -n spark`. Wait a few minutes and check the services output to verify our new service is now having an `EXTERNAL-IP` assigned: 

```bash
$ kubectl get svc -n spark --field-selector metadata.name=spark-pi-loadbalancer
NAME                    TYPE           CLUSTER-IP   EXTERNAL-IP                                                               PORT(S)        AGE
spark-pi-loadbalancer   LoadBalancer   10.0.3.19    a55f8bba6615346149d96bf438d87438-1803869262.us-west-2.elb.amazonaws.com   80:31917/TCP   10m
```

Now you can use `a55f8bba6615346149d96bf438d87438-1803869262.us-west-2.elb.amazonaws.com` address in your browser to access Spark UI:

![](./resources/img/spark-ui-2.png)

> Spark UI provides a valuable insights into various aspects of an application, but it's main drawback is that it's only available for the duration of the application. If you want to be able to explore an application's UI after it's finished, please consider setting up the [Spark History Server](./history-server.md).

Reviewer: ```suggestion
Spark UI provides valuable insights into various aspects of an application, but its main drawback is that it's only available for the duration of the application. If you want to be able to explore an application's UI after it's finished, consider setting up the [Spark History Server](./history-server.md).
```

- kubectl 
- KUDO Spark installed (please refer to the [Installation](installation.md))

In order to deploy a Spark Application to Kubernetes using the KUDO Spark Operator, it should be described as a Kubernetes object. To do that, create a specification in `yaml` format with all the necessary configuration required for the application.

Let's take a simple `SparkPi` application as an example. The `yaml` specification could be found here: [spark-pi.yaml](resources/spark-pi.yaml)

```yaml
apiVersion: "sparkoperator.k8s.io/v1beta2"
kind: SparkApplication
metadata:
  name: spark-pi
  namespace: spark
spec:
  type: Scala
  mode: cluster
  image: "mesosphere/spark:spark-2.4.3-hadoop-2.9-k8s"
  imagePullPolicy: Always
  mainClass: org.apache.spark.examples.SparkPi
  mainApplicationFile: "local:///opt/spark/examples/jars/spark-examples_2.11-2.4.3.jar"
  arguments:
    - "150000"
  sparkConf:
    "spark.ui.port": "4041"
  sparkVersion: "2.4.3"
  restartPolicy:
    type: Never
  driver:
    cores: 1
    memory: "512m"
    labels:
      version: 2.4.3
    serviceAccount: spark-driver
  executor:
    cores: 1
    instances: 2
    memory: "512m"
    labels:
      version: 2.4.3
```

> Basically, all the Spark application configuration is placed under `spec` section. Here you can specify Spark related configuration properties, such as number of executors, number of cores for drivers/executors, amount of memory and the other. There is also a `sparkConf` section, where you can place configuration parameters in a form of key-value pairs. In the example we override the default `spark.ui.port` with a custom value.

Reviewer 2: ```suggestion
Basically, all the Spark application configuration is placed under `spec` section. Here you can specify Spark related configuration properties, such as number of executors, number of cores for drivers/executors, amount of memory, etc. There is also a `sparkConf` section, where you can place configuration parameters in the form of key-value pairs. In this example, we override the default `spark.ui.port` with a custom value.
```

spark-pi   1m
```

In the example above `-n` flag specifies the namespace where the application is deployed. 

To describe the `spark-pi` application, use the following command:
```bash
$ kubectl describe sparkapplications spark-pi -n spark
```

To get the list of pods:
```bash
$ kubectl get pods -n spark
NAME                                  READY   STATUS      RESTARTS   AGE
spark-gq4bf8-spark-867b8c8d79-btt4q   1/1     Running     0          58m
spark-gq4bf8-spark-init-cdcrn         0/1     Completed   0          58m
spark-pi-1571911449587-exec-1         1/1     Running     0          4s
spark-pi-1571911449587-exec-2         1/1     Running     0          4s
spark-pi-driver                       1/1     Running     0          11s
```
From the output above we can see the application has been created and is currently running as three pods: one driver and two executors.

To see pod's logs, use:
```
$ kubectl logs -f spark-pi-driver -n spark
```
Get the last `n` lines of pod's log:
```
$ kubectl logs --tail=n spark-pi-driver -n spark
```
You can get the full description of the driver pod with the following command: 
```
$ kubectl describe pod spark-pi-driver -n spark
```

> Verify the driver pod is completed successfully:

Reviewer 2: ```suggestion
Verify the driver pod has completed successfully:
```

```

Specifying `spark-role: driver` as a `selector` is basically the mechanism for our service to target the required pod.

Create the service using `kubectl create -f ./resources/spark-pi-service.yaml -n spark`. Wait a few minutes and check the services output to verify our new service is now having an `EXTERNAL-IP` assigned: 

```bash
$ kubectl get svc -n spark --field-selector metadata.name=spark-pi-loadbalancer
NAME                    TYPE           CLUSTER-IP   EXTERNAL-IP                                                               PORT(S)        AGE
spark-pi-loadbalancer   LoadBalancer   10.0.3.19    a55f8bba6615346149d96bf438d87438-1803869262.us-west-2.elb.amazonaws.com   80:31917/TCP   10m
```

Now you can use `a55f8bba6615346149d96bf438d87438-1803869262.us-west-2.elb.amazonaws.com` address in your browser to access Spark UI:

![](./resources/img/spark-ui-2.png)

Spark UI provides valuable insights into various aspects of an application, but its main drawback is that it's only available for the duration of the application. If you want to be able to explore an application's UI after it's finished, consider setting up the [Spark History Server](./history-server.md).

> You also might want to consider setting up a complete monitoring solution to be able to visualize and analyze your Spark cluster performance using real-time metrics and dashboards. To do so, please refer to the [Monitoring](monitoring.md) section of the documentation. 

Reviewer 2: Excellent footnotes pointing to other docs that carry users to the next step
